YesterdayApi/Core/Identity/IdentityService.cs

```diff
@@ -0,0 +1,125 @@
+using System.Net.Http;
+using System.Threading.Tasks;
+using AutoMapper;
+using IdentityModel.Client;
+using IdentityServer.Core.Users;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.Extensions.Configuration;
+using YesterdayApi.Core.Identity.Web;
+using YesterdayApi.Core.Users;
+using YesterdayApi.Utilities.Exceptions;
+
+namespace YesterdayApi.Core.Identity
+{
+    public class IdentityService : IIdentityService
+    {
+        private const string DiscoveryUrl = "http://localhost:5000";
+        private readonly IUserIdentityService _userIdentityService;
+        private readonly ITokenDecoder _tokenDecoder;
+        private readonly IMapper _mapper;
+        private readonly IConfiguration _configuration;
+        private readonly HttpClient _client;
+
+        public IdentityService(IUserIdentityService userIdentityService, ITokenDecoder tokenDecoder, IMapper mapper, IConfiguration configuration)
+        {
+            _userIdentityService = userIdentityService;
+            _tokenDecoder = tokenDecoder;
+            _mapper = mapper;
+            _configuration = configuration;
+            _client = new HttpClient();
+        }
+
+        public async Task<string> GetAccessToken(UserIdentityRequest userIdentityRequest)
+        {
+            var disco = await GetDiscoveryDocument();
+
+            var tokenResponse = await RequestPasswordToken(disco.TokenEndpoint, userIdentityRequest);
+
+            if (tokenResponse.IsError)
+            {
+                throw new BadRequestException("Error", tokenResponse.Error);
+            }
+
+            await _userIdentityService.UpdateUserRefreshToken(userIdentityRequest.UserName, tokenResponse.RefreshToken);
+            return tokenResponse.AccessToken;
+        }
+
+        public async Task<string> RefreshAccessToken(string authHeader)
+        {
+            var disco = await GetDiscoveryDocument();
+
+            var id = _tokenDecoder.GetUserId(authHeader);
+
+            var user = await _userIdentityService.GetById(id);
+
+            var tokenResponse = await RequestRefreshToken(disco.TokenEndpoint, user.RefreshToken);
+
+            if (tokenResponse.IsError)
+            {
+                throw new BadRequestException("Error", tokenResponse.Error);
+            }
+
+            return tokenResponse.AccessToken;
+        }
+
+        public async Task<IdentityResult> RegisterUser(UserRegistrationRequest userRegistrationRequest)
+        {
+            var newUser = _mapper.Map<User>(userRegistrationRequest);
+            var result = await _userIdentityService.CreateAsync(newUser, userRegistrationRequest.Password);
+
+            return result;
+        }
+
+        public async Task SetAdmin(string userName)
```

Comment:
Why do you prefer using username instead of id? Id is already indexed by db so it's much faster.

Reply:
How would an admin know someone else's id? I feel like username is more accessible to the admin, so I chose to use it, but if you provide a method of finding out a user's id, I'll gladly change it 👍

Reply:
Admin almost always need to get users (like list them) before setting them admin privileges - through some ui. Otherwise we'll still do this from sql query. But whatever 😇